fdbserver/RestoreMaster.actor.h

```diff
@@ -90,98 +101,188 @@ struct RestoreMasterData : RestoreRoleData, public ReferenceCounted<RestoreMaste
-			for (auto& rangeFile : versionBatch->second.rangeFiles) {
-				rangeFile.fileIndex = ++fileIndex;
-				TraceEvent("FastRestore")
-				    .detail("VersionBatchId", versionBatchId)
-				    .detail("RangeFile", rangeFile.toString());
+		int logIdx = 0;
+		std::vector<RestoreFileFR> retLogs;
+		// Scan all logFiles every time to avoid assumption on log files' version ranges.
+		// For example, we do not assume each version range only exists in one log file
+		while (logIdx < logFiles.size()) {
+			Version begin = std::max(prevVersion, logFiles[logIdx].beginVersion);
+			Version end = std::min(nextVersion, logFiles[logIdx].endVersion);
+			if (begin < end) { // logIdx file overlap in [prevVersion, nextVersion)
+				double ratio = (end - begin) / (logFiles[logIdx].endVersion - logFiles[logIdx].beginVersion);
+				size += logFiles[logIdx].fileSize * ratio;
+				retLogs.push_back(logFiles[logIdx]);
 			}
-			versionBatchId++;
+			++logIdx;
 		}
+		return std::make_tuple(size, rangeIdx, retLogs);
+	}
 
-		TraceEvent("FastRestore").detail("VersionBatches", versionBatches->size());
-		// Sanity check
-		std::set<uint32_t> fIndexSet;
-		for (auto& versionBatch : *versionBatches) {
-			Version prevVersion = 0;
-			for (auto& logFile : versionBatch.second.logFiles) {
-				TraceEvent("FastRestore").detail("PrevVersion", prevVersion).detail("LogFile", logFile.toString());
-				ASSERT(logFile.beginVersion >= versionBatch.second.beginVersion);
-				ASSERT(logFile.endVersion <= versionBatch.second.endVersion);
-				ASSERT(prevVersion <= logFile.beginVersion);
-				prevVersion = logFile.endVersion;
-				ASSERT(fIndexSet.find(logFile.fileIndex) == fIndexSet.end());
-				fIndexSet.insert(logFile.fileIndex);
+	// Split backup files into version batches, each of which has similar data size
+	// Input: sorted range files, sorted log files;
+	// Output: a set of version batches whose size is less than opConfig.batchSizeThreshold
+	//    	   and each mutation in backup files is included in the version batches exactly once.
+	// Assumption 1: input files has no empty files;
+	// Assumption 2: range files at one version <= batchSizeThreshold.
+	void buildVersionBatches(const std::vector<RestoreFileFR>& rangeFiles, const std::vector<RestoreFileFR>& logFiles,
+	                         std::map<Version, VersionBatch>* versionBatches) {
+		bool rewriteNextVersion = false;
+		int rangeIdx = 0;
+		int logIdx = 0; // Ensure each log file is included in version batch
+		Version prevEndVersion = 0;
+		Version nextVersion = 0; // Used to calculate the batch's endVersion
+		VersionBatch vb;
+		vb.beginVersion = 0; // Version batch range [beginVersion, endVersion)
+
+		while (rangeIdx < rangeFiles.size() || logIdx < logFiles.size()) {
+			if (!rewriteNextVersion) {
+				if (rangeIdx < rangeFiles.size() && logIdx < logFiles.size()) {
+					// nextVersion as endVersion is exclusive in the version range
+					nextVersion = std::max(rangeFiles[rangeIdx].version + 1, nextVersion);
```

Comment:
I'm not certain, but it's possible this max isn't necessary.

Reply:
This is because of the example I showed in the above comment. In that case, `nextVersion` will be 1 version larger than the file's version.